README.md
torrentbox Cookbook [![Build Status](https://travis-ci.org/haganenorekenjutsushi/torrentbox.svg?branch=master)](https://travis-ci.org/haganenorekenjutsushi/torrentbox)
===================
Orchestration cookbook for combination MediaCentre/Torrentbox servers in home environment.


Requirements
------------

#### packages
- `apache2` - torrentbox needs apache to display the dashboard
- `sickrage` - torrentbox needs sickrage to organise TV shows
- `couchpotato` - torrentbox needs couchpotato to organise Movies
- `transmission` - torrentbox needs transmission to download torrents
- `snapraid` - torrentbox needs apache to create disk parities
- `mhddfs` - torrentbox needs mhddfs to pool disks
- `cifs-utils` - torrentbox needs cifs to create shares

Attributes
----------


#### torrentbox::default
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['torrentbox']['dashboard']['path']</tt></td>
    <td>string</td>
    <td>location of dashboard website</td>
    <td><tt>/var/www/html/</tt></td>
  </tr>
   <tr>
    <td><tt>['torrentbox']['dashboard']['repo']</tt></td>
    <td>string</td>
    <td>git url of dashboard python repo</td>
    <td><tt>git://github.com/haganenorekenjutsushi/torrentbox-dashboard.git</tt></td>
  </tr>
  <tr>
    <td><tt>['torrentbox']['cifsshares']</tt></td>
    <td>array</td>
    <td>array of remote cifs shares to mount  (and their local/remote paths)</td>
    <td><tt>null</tt></td>
  </tr>
  <tr>
    <td><tt>['torrentbox']['disks']</tt></td>
    <td>array</td>
    <td>disks and their options</td>
    <td><tt>null</tt></td>
  </tr>
  <tr>
    <td><tt>['torrentbox']['cifscredentials']['user']</tt></td>
    <td>string</td>
    <td>name of the user to populate .smbcredentials with</td>
    <td><tt>downloadmanager</tt></td>
  </tr>
   <tr>
    <td><tt>['torrentbox']['cifscredentials']['password']</tt></td>
    <td>string</td>
    <td>password of the user to populate .smbcredentials with</td>
    <td><tt>P@sswort</tt></td>
  </tr>
</table>

Usage
-----
#### torrentbox::default

```json
{
  "run_list": [
    "recipe[torrentbox]",
    "recipe[torrentbox-sickrage]",
    "recipe[transmission]",
    "recipe[torrentbox-couchpotatoserver]"
  ],
  "torrentbox": {
   
    "disks": [
      {
        "device" : "cb92d746-cb71-491e-8f3a-86eef42ce878",
        "device_type": "uuid",
        "fstype" : "ext4",
        "path" : "/media/Parity1",
        "options" : "defaults"
      }
    ]
  }
}
```

Contributing
------------

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github




CHANGELOG.md
torrentbox CHANGELOG
====================

This file is used to list changes made in each version of the torrentbox cookbook.

0.1.0
-----
- [your_name] - Initial release of torrentbox

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


test/fixtures/cookbooks/test/metadata.rb
name 'test'
maintainer 'The Authors'
maintainer_email 'you@example.com'
license 'all_rights'
description 'Installs/Configures test'
long_description 'Installs/Configures test'
version '0.1.0'


test/fixtures/cookbooks/test/recipes/default.rb
bash 'Setup Disks' do
  user 'root'
  cwd '/tmp'
  code <<-EOH
  #!/bin/bash

echo "Setting up test disks..."
{
  # Unmount all disks in case they're already mounted (surpress exit code as they won't be the first time!)
  sudo umount /media/disk1 || true
  sudo umount /media/disk2 || true
  sudo umount /media/disk3 || true

  # Partition all three new disks
  (echo o; echo n; echo p; echo 1; echo ; echo; echo w) | sudo fdisk /dev/sdb
  (echo o; echo n; echo p; echo 1; echo ; echo; echo w) | sudo fdisk /dev/sdc
  (echo o; echo n; echo p; echo 1; echo ; echo; echo w) | sudo fdisk /dev/sdd

  # Format all three new disks
  sudo mkfs.ext4 /dev/sdb1  -E lazy_itable_init=1
  sudo mkfs.ext4 /dev/sdc1  -E lazy_itable_init=1
  sudo mkfs.ext4 /dev/sdd1  -E lazy_itable_init=1

  # Create mount points for all three new disks (surpress exit code as they may already exist)
  sudo mkdir /media/disk1 || true
  sudo mkdir /media/disk2 || true
  sudo mkdir /media/disk3 || true

  # Mount all three new disks
  sudo mount -t ext4 /dev/sdb1 /media/disk1
  sudo mount -t ext4 /dev/sdc1 /media/disk2
  sudo mount -t ext4 /dev/sdd1 /media/disk3

  # Create some files
  sudo dd if=/dev/urandom of=/media/disk1/bigfile1.not bs=1M count=2
  sudo dd if=/dev/urandom of=/media/disk1/bigfile2.not bs=1M count=2
  sudo dd if=/dev/urandom of=/media/disk1/bigfile3.not bs=1M count=2
} &> /dev/null
echo "Disk setup complete!"
  EOH
  not_if "grep \"/dev/sdb1 /media/disk1 ext4 rw 0 0\" /etc/mtab"
end


test/integration/default/serverspec/default_spec.rb
require 'spec_helper'

describe port(80) do
  it { should be_listening }
end

describe file('/var/www/html/index.py') do
  it { should be_file }
  it { should be_executable }
end

describe service('apache2') do
  it { should be_enabled   }
  it { should be_running   }
end

describe service('docker') do
  it { should be_enabled   }
  it { should be_running   }
end

# SickRage
describe port(8081) do
  it { should be_listening }
end

# Couchpotato
describe port(5050) do
  it { should be_listening }
end

# File sharing
describe port(445) do
  it { should be_listening }
end

# Plex
describe port(32_400) do
  it { should be_listening }
end

# Transmission
describe port(9091) do
  it { should be_listening }
end

# Check packages installed
%w(python).each do |package|
  describe package(package) do
    it { should be_installed }
  end
end


recipes/default.rb
#
# Cookbook Name:: torrentbox
# Recipe:: default
#
# Copyright 2014, YOUR_COMPANY_NAME
#
# All rights reserved - Do Not Redistribute
#
include_recipe 'apt::default'
include_recipe 'git::default'

# Mount Disks
node['torrentbox']['disks'].each do |disk|
  directory disk['path'] do
    action :create
  end

  log "Mounting #{disk['path']}"
  mount disk['path'] do
    device disk['device']
    device_type disk['device_type']
    fstype disk['fstype']
    options disk['options']
    action [:mount, :enable]
    not_if { ::File.directory?('/vagrant') }
  end
end

# Setup mhddfs
include_recipe 'torrentbox::mhddfs'

# Install Dashboard
include_recipe 'torrentbox::dashboard'

# Install SickRage, CouchPotato
include_recipe 'torrentbox::mediamanagement'

############################################
## Setup fileshare user
############################################

user 'DownloadManager' do
  comment 'Samba Test User'
  home '/home/DownloadManager'
  shell '/bin/bash'
end

include_recipe 'samba::server'


recipes/dashboard.rb

# Prevent "unable to execute 'x86_64-linux-gnu-gcc': No such file or directory" error when installing Python packages
package 'build-essential' do
  action :install
end

# Install Python
include_recipe 'python::default'

# Install Apache2
include_recipe 'apache2::default'
include_recipe 'apache2::mod_cgid'

# Install required python modules
python_pip 'psutil'

# Configure default website
web_app '001-default' do
  template 'dashboard-site.conf.erb'
  server_name '*'
  cookbook 'torrentbox'
end

# Delete the default page so we can cleanly bring the dashboard in
file "#{node['torrentbox']['dashboard']['path']}/index.html" do
  action :delete
end

# Control the dashboard dir
directory node['torrentbox']['dashboard']['path'] do
  mode '0755'
  action :create
end

log 'Deploying dashboard'
git node['torrentbox']['dashboard']['path'] do
  repository node['torrentbox']['dashboard']['repo']
  notifies :restart, 'service[apache2]', :delayed
end

# Configure the dashboard
template "#{node['torrentbox']['dashboard']['path']}/config.py" do
  source 'dashboard.py.erb'
  variables(
    diskUsagePath: node['torrentbox']['dashboard']['diskUsagePath']
  )
  mode '777'
  action :create
end

# Ensure we can execute the dashboard's index page
file "#{node['torrentbox']['dashboard']['path']}/index.py" do
  mode '777'
end


recipes/mediamanagement.rb
# Install Docker
include_recipe 'docker::default'

# Create directories
[
  node['torrentbox']['directories']['tv_downloads'], node['torrentbox']['directories']['movie_downloads'],
  node['torrentbox']['directories']['incomplete_downloads'],
  node['torrentbox']['directories']['tv'], node['torrentbox']['directories']['movies'],
  '/opt/CouchPotato', '/opt/SickRage', '/opt/plex/Library/Application Support/Plex Media Server/', '/opt/transmission'
].each do |share|
  directory share do
    owner 'root'
    action :create
    recursive true
  end
end

# Default transmission settings
cookbook_file 'settings.json' do
  path '/opt/transmission/settings.json'
  action :create_if_missing
end

# Default Plex settings
cookbook_file 'Preferences.xml' do
  path "/opt/plex/Library/Application\ Support/Plex\ Media\ Server/Preferences.xml"
  action :create_if_missing
end

# Default Couchpotato settings
cookbook_file 'CouchPotato.cfg' do
  path '/opt/CouchPotato/CouchPotato.cfg'
  action :create_if_missing
end

# Default SickRage settings
cookbook_file 'config.ini' do
  path '/opt/SickRage/config.ini'
  action :create_if_missing
end

# Pull latest image
docker_image 'timhaak/sickrage' do
  action :pull_if_missing
end
# Pull latest image
docker_image 'timhaak/transmission' do
  action :pull_if_missing
end
# Pull latest image
docker_image 'timhaak/couchpotato' do
  action :pull_if_missing
end
# Pull latest image
docker_image 'timhaak/plex' do
  cmd_timeout 900
  action :pull_if_missing
end

# Install Transmission
docker_container 'dperson/transmission' do
  container_name 'transmission'
  detach true
  cmd_timeout 900
  port %w('9091:9091' '45555:45555' '51415:51415' '51415:51415/udp')
  env %w('TRUSER=username' 'TRPASSWD=password')
  volume %W('/opt/transmission:/config'
            "#{node['torrentbox']['directories']['tv_downloads'].gsub " ","\\ "}:/tv_downloads"
            "#{node['torrentbox']['directories']['movie_downloads'].gsub " ","\\ "}:/movie_downloads"
            "#{node['torrentbox']['directories']['incomplete_downloads'].gsub " ","\\ "}:/var/lib/transmission-daemon/incomplete"
            "/opt/transmission:/var/lib/transmission-daemon/info"
            '/etc/localtime:/etc/localtime:ro')
end

# Install SickRage
docker_container 'timhaak/sickrage' do
  container_name 'sickrage'
  detach true
  port '8081:8081'
  link 'transmission:transmission'
  volume %W('/opt/SickRage:/config'
            "#{node['torrentbox']['directories']['tv_downloads'].gsub " ","\\ "}:/tv_downloads"
            "#{node['torrentbox']['directories']['tv'].gsub " ","\\ "}:/tv"
            '/etc/localtime:/etc/localtime:ro')
end

# Install CouchPotato
docker_container 'timhaak/couchpotato' do
  container_name 'couchpotato'
  detach true
  port '5050:5050'
  link 'transmission:transmission'
  volume %W('/opt/CouchPotato:/config'
            "#{node['torrentbox']['directories']['movie_downloads'].gsub " ","\\ "}:/movie_downloads"
            "#{node['torrentbox']['directories']['movies'].gsub " ","\\ "}:/movies"
            '/etc/localtime:/etc/localtime:ro')
end

# Install Plex
docker_container 'timhaak/plex' do
  container_name 'plex'
  detach true
  port '32400:32400'
  volume %W('/opt/plex:/config'
            "#{node['torrentbox']['directories']['tv'].gsub " ","\\ "}:/tv"
            "#{node['torrentbox']['directories']['movies'].gsub " ","\\ "}:/movies"
            '/etc/localtime:/etc/localtime:ro')
end
